sonar-html-plugin/src/main/java/org/sonar/plugins/html/api/accessibility/ControlGroup.java

```diff
@@ -0,0 +1,138 @@
+import org.sonar.plugins.html.node.TagNode;
+
+public class ControlGroup {
+  public static boolean belongsToAutofillExpectationMantleControlGroup(TagNode node) {
+    if (!node.getNodeName().equals(HtmlConstants.NAME_INPUT)) {
+      return true;
+    }
+
+    String type = node.getAttribute("type");
+
+    return type == null || !type.equals(HtmlConstants.TYPE_HIDDEN);
+  }
+
+  public static boolean belongsToDateControlGroup(TagNode node) {
+    if (belongsToTextControlGroup(node)) {
+      return true;
+    }
+
+    String type = node.getAttribute("type");
+
+    return type != null && type.equals("date");
+  }
+
+  public static boolean belongsToMonthControlGroup(TagNode node) {
+    String type = node.getAttribute("type");
+
+    if (type != null && type.equals("month")) {
+      return true;
+    }
+
+    return belongsToTextControlGroup(node);
+  }
+
+  public static boolean belongsToMultilineControlGroup(TagNode node) {
+    if (node.getNodeName().equals(HtmlConstants.NAME_TEXTAREA) || node.getNodeName().equals("select")) {
+      return true;
+    }
+
+    if (!node.getNodeName().equals(HtmlConstants.NAME_INPUT)) {
+      return false;
+    }
+
+    String type = node.getAttribute("type");
+
+    return type != null && type.equals(HtmlConstants.TYPE_HIDDEN);
+  }
+
+  public static boolean belongsToNumericControlGroup(TagNode node) {
+    String type = node.getAttribute("type");
+
+    if (type != null && type.equals("number")) {
+      return true;
+    }
+
+    return belongsToTextControlGroup(node);
+  }
+
+  public static boolean belongsToPasswordControlGroup(TagNode node) {
+    String type = node.getAttribute("type");
+
+    if (type != null && type.equals("password")) {
+      return true;
+    }
+
+    return belongsToTextControlGroup(node);
+  }
+
+  public static boolean belongsToTelControlGroup(TagNode node) {
+    String type = node.getAttribute("type");
+
+    if (type != null && type.equals("tel")) {
+      return true;
+    }
+
+    return belongsToTextControlGroup(node);
+  }
+
+  public static boolean belongsToTextControlGroup(TagNode node) {
+    if (node.getNodeName().equals(HtmlConstants.NAME_TEXTAREA) || node.getNodeName().equals("select")) {
+      return true;
+    }
+
+    if (!node.getNodeName().equals(HtmlConstants.NAME_INPUT)) {
+      return false;
+    }
+
+    String type = node.getAttribute("type");
+
+    if (type == null) {
+      return false;
+    }
+
+    return type.equals(HtmlConstants.TYPE_HIDDEN) || type.equals("text") || type.equals("search");
+  }
+
+  public static boolean belongsToUrlControlGroup(TagNode node) {
+    String type = node.getAttribute("type");
+
+    if (type != null && type.equals("url")) {
+      return true;
+    }
+
+    return belongsToTextControlGroup(node);
+  }
+
+  public static boolean belongsToUsernameControlGroup(TagNode node) {
+    String type = node.getAttribute("type");
+
+    if (type != null && type.equals("email")) {
+      return true;
+    }
```

Comment:
After seeing this pattern multiple times, I think it would make sense to extract it into a method to `TagNode`:

```java
class TagNode {
  // ...
  public boolean hasAttributeValues(String attributeName, String... values) {
    var attributeValue = getAttribute(attributeName);
    if (attributeValue != null) {
      for (var value : values) {
        if (attributeValue.equalsIgnoreCase(value)) {
          return true;
        }
      }
    }
    return false;
  }
}
```

To be done in a different PR and (maybe go over all the rules that could benefit from this method.)

Reply:
Definitely!